### PR TITLE
Getting started setup.py script

### PR DESCRIPTION
**Pull request by boxed**
Also improves experience under Visual Studio (for Mac at least).

This is a bit of a rough draft obviously. The `setup.py` file only works for macOS because it assumes homebrew exists, and python isn't even installed by default on windows so that's a problem. 

But I think something like this, but built out to support apt-get/yum/whatever and windows would be a lot nicer when getting started.

I'd love to get your feedback on this.

### CODE REVIEW

**Comment by boxed** [comment]
Oh yea... I tried to write the setup script in F#, but I couldn't figure that out. Seems like there's no shebang for .fsx files? And I couldn't understand how to execute shell commands either :P

**Comment by alfonsogarciacaro** [comment]
Thanks @boxed, I think this is very good. It'd be nice if it could be adapted to other platforms (there're many Windows machines without Python installed). Using an .fsx script may be a bit cumbersome as you need to have `fsi` in PATH and usually also [FAKE](http://fsharp.github.io/FAKE/) helpers, but maybe we can use a JS script like the one in [fable-powerpack]() instead.

Also, fyi we will probably create soon a template to take advantage of the new `new` (pun intended) command in [dotnet sdk](https://github.com/dotnet/core/blob/master/release-notes/1.0/1.0.3-SDK-RC4.md#dotnet-new).

Anyway, I'll let @Pauan take a look at this before anything :)

File: GettingStarted.sln [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (Pauan) [inline GettingStarted.sln]: Is this file actually needed for Visual Studio on Mac?

File: setup.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import json
+import shutil
+
+check_call('brew install npm', shell=True)
+check_call('brew install yarn', shell=True)
+check_call('npm install --global yarn', shell=True)
+check_call('yarn install', shell=True)
+
+# update GUID
+new_guid = check_output('uuidgen').strip()
```
Comment (Pauan) [inline setup.py]: According to Don Syme we don't actually need a new GUID, so we can remove this.

File: setup.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    fsproj = f.read()
+with open('src/Main.fsproj', 'w') as f:
+    f.write(fsproj.replace('4057826A-1B4A-4553-832E-9570A3394346', new_guid))
+
+package = json.loads(open('package.json').read())
+package['name'] = raw_input('Project name: ')
+package['description'] = raw_input('Project description: ')
+package['repository'] = raw_input('Project repository: ')
+with open('package.json', 'w') as f: 
+    f.write(json.dumps(package, indent=4))
```
Comment (Pauan) [inline setup.py]: I prefer 2 indent rather than 4. But since F# code uses 4 indent, I'll need to think about whether 2 or 4 is better.

File: GettingStarted.sln [hunk 4]
```diff
@@ -0,0 +1,17 @@
+
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 2012
+Project("{f2a71f9b-5d33-465a-a702-920d77279786}") = "Main", "src\Main.fsproj", "{968747A0-B0E6-4612-808B-89EDEC1DEFB8}"
```
Comment (Pauan) [inline GettingStarted.sln]: I'm guessing these GUIDs were auto-generated by Visual Studio. I wonder if that's portable, or whether we need to adjust the GUIDs per project.

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    
+
+shutil.move('GettingStarted.sln', package['name'] + '.sln')
+
+print 'Done! Now use "yarn" or "yarn watch" to compile.'
```
Comment (Pauan) [inline setup.py]: This is redundant since line 10 already uses `yarn install`

In fact, it may be better to move all of the `check_call` functions to the end, so that way `yarn install` will correctly pick up the changes to `GettingStarted.sln`

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
+
+from subprocess import check_call, check_output
+import json
+import shutil
+
+check_call('brew install npm', shell=True)
+check_call('brew install yarn', shell=True)
+check_call('npm install --global yarn', shell=True)
```
Comment (Pauan) [inline setup.py]: This seems redundant, since the above line uses `brew install yarn` which should be good enough.

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env python
+
+from subprocess import check_call, check_output
+import json
+import shutil
+
+check_call('brew install npm', shell=True)
```
Comment (Pauan) [inline setup.py]: I don't think this is necessary, I think `yarn` alone should be enough.

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+check_call('brew install yarn', shell=True)
```
Comment (Pauan) [inline setup.py]: The [Yarn docs](https://yarnpkg.com/en/docs/install#mac-tab) suggest using `brew update`, so that should probably be added.

It also suggests changing the `PATH`, but that may be too complicated for `setup.py`

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+package['description'] = raw_input('Project description: ')
+package['repository'] = raw_input('Project repository: ')
+with open('package.json', 'w') as f: 
+    f.write(json.dumps(package, indent=4))
+    
+
+shutil.move('GettingStarted.sln', package['name'] + '.sln')
+
+print 'Done! Now use "yarn" or "yarn watch" to compile.'
+print 'The output file is in dist/index.html'
```
Comment (Pauan) [inline setup.py]: The output is actually in `dist/umd/Main.js`, so I would prefer if it was worded as `You can open dist/index.html in your browser to run your code`

File: src/Main.fsproj [hunk 10]
```diff
@@ -14,8 +14,12 @@
    !  They must start with ./ or ../ so Fable knows that they're relative paths.
    !-->
   <ItemGroup>
-    <Reference Include="../node_modules/fable-core/Fable.Core.dll" />
-    <Reference Include="../node_modules/fable-powerpack/Fable.PowerPack.dll" />
+    <Reference Include="Fable.Core">
+      <HintPath>..\node_modules\fable-core\Fable.Core.dll</HintPath>
```
Comment (Pauan) [inline src/Main.fsproj]: Is the `HintPath` needed for Visual Studio on Mac?

Also, are the `\` needed, or does `/` work as well?

I'm a Linux guy, so I'm probably biased, but I prefer `/` over `\`

File: src/Main.fsproj [hunk 11]
```diff
@@ -48,6 +52,9 @@
     <Reference Include="System.Core" />
     <Reference Include="System.Numerics" />
   </ItemGroup>
+  <ItemGroup>
+    <None Include="js\Message.js" />
```
Comment (Pauan) [inline src/Main.fsproj]: Is this needed for Visual Studio on Mac?

File: GettingStarted.sln (hunk 1 above)
Comment (boxed) [inline GettingStarted.sln]: I don't know... if I open src/Main.fsproj it creates src/Main.sln. That seems wrong to me, a sln file in the right place seems better. Note that the python setup script renames it properly.

File: setup.py (hunk 8 above)
Comment (boxed) [inline setup.py]: The path thing is just wrong. I have access to yarn globally. Updating brew seems semi-automatic so probably not needed either, unless you want to be 100% certain you're running the latest version.

File: setup.py (hunk 6 above)
Comment (boxed) [inline setup.py]: Ah, my bad, left over from older versions.

File: setup.py (hunk 2 above)
Comment (boxed) [inline setup.py]: It says "very soon", not "now" on that page?

File: setup.py (hunk 3 above)
Comment (boxed) [inline setup.py]: Crazy javascript people seem to be the only ones holding out on 2 :P

File: setup.py (hunk 5 above)
Comment (boxed) [inline setup.py]: Seems this comment was made in the wrong place? Moving the sln file and printing a nice helpful message doesn't become obsolete because there's a "yarn install" somewhere does it?

File: setup.py (hunk 9 above)
Comment (boxed) [inline setup.py]: Good point.

File: src/Main.fsproj (hunk 10 above)
Comment (boxed) [inline src/Main.fsproj]: I changed this because the old ones didn't work so yea, and the new versions are from clicking around in the GUI. I agree it's super weird with \, I didn't check if / works too. But if you change anything you can be sure it'll get rewritten as \ so I think it's probably better to keep it like that :(

File: src/Main.fsproj (hunk 11 above)
Comment (boxed) [inline src/Main.fsproj]: Well.. I think it's super annoying that a source file included in the project isn't visible in the side bar. "Needed" might be a bit strong a word...

File: setup.py (hunk 2 above)
Comment (Pauan) [inline setup.py]: There's two separate issues:

1. Don Syme says that right now it will work with the existing GUIDs, so we don't need to generate new GUIDs.

2. Alfonso is saying that soon we'll be able to remove GUIDs completely.

So we should be able to remove the *new GUID generation* right now, but the existing GUIDs have to stay for now.

File: setup.py (hunk 3 above)
Comment (Pauan) [inline setup.py]: And Lisp, and Haskell, and Ruby, and... :stuck_out_tongue: 

2 spaces also isn't standard in JS (Douglas Crockford recommends 4 spaces). But there's a lot of disagreement between different teams.

File: GettingStarted.sln (hunk 1 above)
Comment (Pauan) [inline GettingStarted.sln]: Hm, I'm still not sold on the `sln` file, especially since I think they'll be obsolete (?) soon.

File: setup.py (hunk 8 above)
Comment (Pauan) [inline setup.py]: Do you have access to `yarn` even if you install it through `brew` and **don't** install it with `npm --global`?

File: setup.py (hunk 5 above)
Comment (Pauan) [inline setup.py]: The message implies that you need to run `yarn` after running `setup.py`, but that's redundant since `setup.py` already runs `yarn install`

The README explains both the `yarn` and `yarn watch` commands, so I don't think this message is necessary.

File: src/Main.fsproj (hunk 10 above)
Comment (Pauan) [inline src/Main.fsproj]: That's weird that Visual Studio doesn't support the `Include` property, but oh well, if it's needed then it's needed.

File: src/Main.fsproj (hunk 11 above)
Comment (Pauan) [inline src/Main.fsproj]: I see. In that case I'm okay with this.

File: GettingStarted.sln (hunk 1 above)
Comment (boxed) [inline GettingStarted.sln]: I tried googling but I can't find any reference to solution files becoming obsolete. Do you have a source?

File: GettingStarted.sln (hunk 4 above)
Comment (boxed) [inline GettingStarted.sln]: Yep, and no idea :P

File: setup.py (hunk 7 above)
Comment (boxed) [inline setup.py]: Check

File: setup.py (hunk 8 above)
Comment (boxed) [inline setup.py]: Yep.

File: setup.py (hunk 2 above)
Comment (boxed) [inline setup.py]: Good. I'll remove this and the text in the readme that says this is needed.

File: setup.py (hunk 3 above)
Comment (boxed) [inline setup.py]: Ah, it's 2 before, so yea, I should keep it at that.

File: setup.py (hunk 5 above)
Comment (boxed) [inline setup.py]: Necessary no, but something along these lines is really nice coming at projects like this from a blank slate. Reading a long readme isn't great.

**Comment by boxed** [comment]
I've pushed the changes I think you wanted. Please have a look.

**Comment by Pauan** [comment]
@boxed I'm very sick right now, but I'll look over the changes after I feel better.

File: src/Main.fsproj (hunk 10 above)
Comment (Pauan) [inline src/Main.fsproj]: You should change `test/Test.fsproj` to use `\` as well.

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+
+package = json.loads(open('package.json').read())
+package['name'] = raw_input('Project name: ')
+package['description'] = raw_input('Project description: ')
+package['repository'] = raw_input('Project repository: ')
+with open('package.json', 'w') as f: 
+    f.write(json.dumps(package, indent=2))
+    
+
+shutil.move('GettingStarted.sln', package['name'] + '.sln')
```
Comment (Pauan) [inline setup.py]: From what you said, it seems VS expects the `.sln` file to be in the same directory as the `.fsproj` file.

I think it would be better to:

1. Move `GettingStarted.sln` to `src/Main.sln`
2. Remove this line from `setup.py`, since it's no longer needed.
3. Remove the `import shutil` line, since it's no longer needed.

File: setup.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+
+print 'Done! The project is now built. Use "yarn watch" to live compile as you save.'
```
Comment (Pauan) [inline setup.py]: I think the output should be:
```
Done! The project is now built.
You can open dist/index.html in your browser to run your code.
Use "yarn" to recompile.
Use "yarn watch" to automatically recompile when you save.
```

File: setup.py (hunk 12 above)
Comment (boxed) [inline setup.py]: Moving the sln file into a folder is pretty bad I think, from a usability perspective. It's pretty much the _only_ file that needs to be in the top level directory if you think about it, because it's the one file you open.

The reason I move it like that is that the solution should have the name of the project, that's the point of the file as far as I understand it.

**Comment by Pauan** [comment]
Well, there's still a couple things I don't like, but I'll go ahead and merge this since we need to make some changes for Fable 1.0 anyways.

Thanks for your help!

**Comment by boxed** [comment]
Great!
